sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/s3/S3FileSystemTest.java

```diff
@@ -93,51 +101,101 @@ public void testGetScheme() {
   }
 
   @Test
-  public void testCopyMultipleParts() throws IOException {
-    S3Options pipelineOptions = s3Options();
-    S3FileSystem s3FileSystem = new S3FileSystem(pipelineOptions);
+  public void testCopy() throws IOException {
+    S3FileSystem s3FileSystem = buildMockedS3FileSystem();
+
+    S3ResourceId sourcePath = S3ResourceId.fromUri("s3://bucket/from");
+    S3ResourceId destinationPath = S3ResourceId.fromUri("s3://bucket/to");
+
+    ObjectMetadata objectMetadata = new ObjectMetadata();
+    objectMetadata.setContentLength(0);
+    when(s3FileSystem
+            .getAmazonS3Client()
+            .getObjectMetadata(sourcePath.getBucket(), sourcePath.getKey()))
+        .thenReturn(objectMetadata);
+
+    s3FileSystem.copy(sourcePath, destinationPath);
+
+    verify(s3FileSystem.getAmazonS3Client(), times(1))
+        .copyObject(argThat(notNullValue(CopyObjectRequest.class)));
+
+    // we simulate a big object >= 5GB so it takes the multiPart path
+    objectMetadata.setContentLength((long) 5 * 1024 * 1024 * 1024);
+    when(s3FileSystem
+            .getAmazonS3Client()
+            .getObjectMetadata(sourcePath.getBucket(), sourcePath.getKey()))
+        .thenReturn(objectMetadata);
+
+    try {
+      s3FileSystem.copy(sourcePath, destinationPath);
+    } catch (NullPointerException e) {
+      // ignore failing unmocked path, this is covered by testMultipartCopy test
+    }
 
-    AmazonS3 mockAmazonS3 = Mockito.mock(AmazonS3.class);
-    s3FileSystem.setAmazonS3Client(mockAmazonS3);
+    verify(s3FileSystem.getAmazonS3Client(), never())
+        .copyObject(argThat(nullValue(CopyObjectRequest.class)));
+  }
+
+  @Test
+  public void testAtomicCopy() throws IOException {
+    S3FileSystem s3FileSystem = buildMockedS3FileSystem();
+
+    S3ResourceId sourcePath = S3ResourceId.fromUri("s3://bucket/from");
+    S3ResourceId destinationPath = S3ResourceId.fromUri("s3://bucket/to");
+
+    CopyObjectResult copyObjectResult = new CopyObjectResult();
+    when(s3FileSystem
+            .getAmazonS3Client()
+            .copyObject(argThat(notNullValue(CopyObjectRequest.class))))
+        .thenReturn(copyObjectResult);
+
+    s3FileSystem.atomicCopy(sourcePath, destinationPath);
+
+    verify(s3FileSystem.getAmazonS3Client(), times(1))
+        .copyObject(argThat(notNullValue(CopyObjectRequest.class)));
+  }
+
+  @Test
+  public void testMultipartCopy() throws IOException {
```

Comment:
This test asserts that only one call to the S3 client is issued. Can we add a test for a bigger object that will be split into several pieces and assert that the number of calls to the S3 client is correct (I guess more than 0) ?

Reply:
This test is exactly the original test that existed before and it invokes the `multipartCopy` method (previously called copy) with a 'big' object with two parts (notice that we don't require a huge >5GB object to test that `multipartCopy` works, Amazon support this even for really small objects), The `multipartCopy` method calls the `amazons3,copyPart` method twice internally so I think this cover the case you mention via:
```
    when(s3FileSystem.getAmazonS3Client().copyPart(argThat(notNullValue(CopyPartRequest.class))))
         .thenReturn(copyPartResult1)
         .thenReturn(copyPartResult2);
```
Notice that the validation at the end of the test:
```
  verify(s3FileSystem.getAmazonS3Client(), never())
        .copyObject(argThat(nullValue(CopyObjectRequest.class)));
```
guarantees that the multipartCopy method calls the end of the copy by doing the `CompleteMultipartUploadRequest`. This guarantees that all the calls to the API were complete (Initialize -> CopyPart* -> Complete).

Reply:
Notice that I prefer not to squash the current two commits because they are totally unrelated.

Reply:
ok fair enough for the multi-part, thanks. 
For the squash I prefer not to leave the first commit as it is just formatting. Can you squash with the other and add in the commit message ", and reformat FileBasedSink" ?

Reply:
LGTM modulo squash